Adding accessibility help for verbose hover

aiday-mar opened the pull request:
Adding accessibility help for verbose hover

meganrogge reviewed (changes requested):
Thanks for working on this. I didn't test this, but it LGTM. I suggest adding the `+/-` actions to the accessible view implementation when they're available.

meganrogge commented on src/vs/editor/contrib/hover/browser/hoverAccessibleViews.ts:
```diff
@@ -2,48 +2,205 @@
+		let descriptionWithKb: string;
+		let descriptionWithoutKb: string;
+		switch (action) {
+			case HoverVerbosityAction.Increase:
```
The accessible view service now resolves keybindings for the help dialogs. This is the case so that we have uniform language when there is no assigned keybinding and because we offer a configure keybinding action for commands with no keybindings. 

Here's what the chat accessibility help does for example:

https://github.com/microsoft/vscode/blob/3f44453bbc6da6fc31b04f8602a96fbcb06e64c7/src/vs/workbench/contrib/chat/browser/actions/chatAccessibilityHelp.ts#L42

aiday-mar replied:
Oh I see thanks for the tip, I will implement that

meganrogge reviewed (approved):
Just one comment, but this looks great. Thank you!

aiday-mar commented:
Hi @meganrogge thank you for the review. I have added the usage of `command<keybinding:{0}>` in the content of the provider in order to display the keybindings. To do this I changed the type of the view and the provider from `view` to `help`.

I was testing the feature and I noticed that the when I would click outside of the accessible view, and make it disappear, the hover would not appear anymore thereafter. I dug around and found that this is because the accessible view does not call the `onClose` method when it disappears through a blur event. In my case the `onClose` method does some cleaning and listener disposing which is needed to reset the normal hover functioning. I looked at where `onClose` is called, and noticed it was called in the `hide` arrow function of the `_render` method of the `accessibleView.ts` file. In particular on line 595, when the editor is blurred, we call `this._contextViewService.hideContextView();` but not `hide` and therefore not `onClose`.

https://github.com/microsoft/vscode/blob/869828c0978004ea059416b1ff4c596bbe64ddfe/src/vs/workbench/contrib/accessibility/browser/accessibleView.ts#L593C1-L597C1

I changed the code a bit there to also called `hide` as follows:

https://github.com/microsoft/vscode/pull/212783/files#diff-57bdd64acf33af25dee4c88c656bf7548ab00a0d9eaebaae5e627ddb299a8516R601-R605

I am wondering if there is a specific reason we only call `this._contextViewService.hideContextView()` on the editor widget blur event and if you think the change of the PR is okay?

meganrogge commented:
@aiday-mar I think your change sounds good. Thanks for the info, I will test today to make sure everything still works as expected.

meganrogge commented:
I am working on the fix

meganrogge commented:
I noticed that these actions are present and enabled in the toolbar even when the actions can't be taken. Maybe a context key would be a good idea so they only show when actionable?

<img width="778" alt="Screenshot 2024-05-23 at 11 43 19 AM" src="https://github.com/microsoft/vscode/assets/29464607/98d3190b-9ea0-46ff-b1bf-8fda2b03c9f9">

meganrogge commented:
Summary of my changes:

- reverted your commit which changed this to a help dialog & removed the kb service and resolution
- provided hover content in the accessible view `(alt+f2)` and provided help content for that `(alt+f1)`
- reapplied your changes which removed the kb service and resolution

meganrogge commented:
I think we additionally want an accessibility help dialog when the hover is focused. I will also add that.

meganrogge commented:
I have added `HoverAccessibilityHelp` in addition to the `HoverAccessibleView` and my testing shows it's all working well now.

1. Focus a hover
2. Hear the hint to open the accessibility help dialog, `alt+f1` takes you there
3. Escape that, hover is focused
4. `alt+f2` to go to accessible view, hear hint about opening help
5. `alt+f1` goes to the help dialog
6. `Escape` takes you back to the accessible view
7. `Escape` focuses the hover

aiday-mar commented:
Hi @Meggan thanks a lot for looking into the PR and making the changes! I tested the changes with an extension which provides verbose markdown hovers, and have made some changes on top. The current behavior is as follows:

https://github.com/microsoft/vscode/assets/61460952/5e0637a6-7c91-4320-969c-d08396552682

aiday-mar commented:
Hi @meganrogge I am thinking of merging this PR this iteration. Would you say this PR is ready for merging?
